# Coding Exercises (Part 5)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas: Intermediate

### Exercise 5: Sorting & Filtering 

In [1]:
# run the cell!
import pandas as pd

In [2]:
# run the cell!
listings = pd.read_csv("listings.csv")

In [3]:
# run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price    Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
1     AA     NYSE              Alcoa Corporation       29.15  5.407810e+09   
2   AABA   NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
3    AAC     NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

   ADR TSO  IPO_Year            Sector  \
0      NaN    1999.0     Capital Goods   
1      NaN    2016.0  Basic Industries   
2      NaN       NaN        Technology   
3      NaN    2014.0       Health Care   
4      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

39. __Sort__ the __listings__ DataFrame by the __"Market_Cap"__ column from __high to low__ and __save__ the new order! __Fill in the gaps__!

In [4]:
listings.sort_values(by = "Market_Cap", ascending = False, inplace= True)

__Inspect__ listings! Which Company has the __highest Market Capitalization__ (Name)?

In [5]:
#run the cell!
listings.head()

Symbol Exchange                   Name  Last_Price    Market_Cap  \
11     AAPL   NASDAQ             Apple Inc.      200.62  9.459795e+11   
4109   MSFT   NASDAQ  Microsoft Corporation      120.19  9.221233e+11   
350    AMZN   NASDAQ       Amazon.com, Inc.     1847.33  9.074138e+11   
2767  GOOGL   NASDAQ          Alphabet Inc.     1206.45  8.387076e+11   
2766   GOOG   NASDAQ          Alphabet Inc.     1202.16  8.357253e+11   

      ADR TSO  IPO_Year             Sector  \
11        NaN    1980.0         Technology   
4109      NaN    1986.0         Technology   
350       NaN    1997.0  Consumer Services   
2767      NaN       NaN         Technology   
2766      NaN    2004.0         Technology   

                                             Industry  
11                             Computer Manufacturing  
4109          Computer Software: Prepackaged Software  
350                    Catalog/Specialty Distribution  
2767  Computer Software: Programming, Data Processing  
2766  Computer Software: Programming, Data Processing

The highest Market Capitalization has ... Apple!

40. __Sort__ listings by __IPO_Year__ (ascending) and __Market_Cap__ (descending) and save the new order. 

In [6]:
listings.sort_values(["IPO_Year", "Market_Cap"], ascending = [True, False], inplace = True)

__Inspect__ listings! From all Companies that were listed in __1983__, which one has the __highest Market Capitalization__?

In [7]:
#run the cell!
listings.head(15)

Symbol Exchange                                    Name  Last_Price  \
294    AMAT   NASDAQ                 Applied Materials, Inc.       42.17   
1439   COKE   NASDAQ            Coca-Cola Consolidated, Inc.      294.70   
6612   WDFC   NASDAQ                           WD-40 Company      163.60   
11     AAPL   NASDAQ                              Apple Inc.      200.62   
3539   KLAC   NASDAQ                  KLA-Tencor Corporation      123.34   
5511   SEIC   NASDAQ                 SEI Investments Company       56.06   
1083   CBRL   NASDAQ  Cracker Barrel Old Country Store, Inc.      152.73   
1542   CSII   NASDAQ            Cardiovascular Systems, Inc.       37.99   
2439   FONR   NASDAQ                       Fonar Corporation       20.45   
1688   DAIO   NASDAQ                    Data I/O Corporation        5.40   
2045   EMCI   NASDAQ                EMC Insurance Group Inc.       31.50   
1548   CSPI   NASDAQ                                CSP Inc.       12.20   
312    AMGN   NASDAQ                              Amgen Inc.      193.89   
4730   PAYX   NASDAQ                           Paychex, Inc.       81.14   
1566   CTAS   NASDAQ                      Cintas Corporation      204.97   

        Market_Cap  ADR TSO  IPO_Year                 Sector  \
294   4.003589e+10      NaN    1972.0             Technology   
1439  2.104584e+09      NaN    1972.0  Consumer Non-Durables   
6612  2.257237e+09      NaN    1973.0       Basic Industries   
11    9.459795e+11      NaN    1980.0             Technology   
3539  2.018632e+10      NaN    1980.0          Capital Goods   
5511  8.615867e+09      NaN    1981.0                Finance   
1083  3.671839e+09      NaN    1981.0      Consumer Services   
1542  1.322928e+09      NaN    1981.0            Health Care   
2439  1.326209e+08      NaN    1981.0            Health Care   
1688  4.506536e+07      NaN    1981.0          Capital Goods   
2045  6.816794e+08      NaN    1982.0                Finance   
1548  4.903490e+07      NaN    1982.0             Technology   
312   1.194264e+11      NaN    1983.0            Health Care   
4730  2.916172e+10      NaN    1983.0      Consumer Services   
1566  2.143389e+10      NaN    1983.0  Consumer Non-Durables   

                                               Industry  
294                                      Semiconductors  
1439                Beverages (Production/Distribution)  
6612                                    Major Chemicals  
11                               Computer Manufacturing  
3539                              Electronic Components  
5511                 Investment Bankers/Brokers/Service  
1083                                        Restaurants  
1542                         Medical/Dental Instruments  
2439  Biotechnology: Electromedical & Electrotherape...  
1688                                Electrical Products  
2045                         Property-Casualty Insurers  
1548                                       EDP Services  
312   Biotechnology: Biological Products (No Diagnos...  
4730                    Diversified Commercial Services  
1566                                            Apparel

... Amgen Inc. has the highest Market Capitalization.

41. Set the column __Symbol__ as new Index of the __listings__ DataFrame! __Save__ the change!

In [8]:
listings.set_index("Symbol", inplace = True)

42. __Sort__ the __listings__ DataFrame by the new __index__ from __low to high__. __Save__ the new order!

In [9]:
listings.sort_index(ascending = True, inplace= True)

In [10]:
# run the cell!
listings.head()

Exchange                           Name  Last_Price    Market_Cap  \
Symbol                                                                     
A          NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
AA         NYSE              Alcoa Corporation       29.15  5.407810e+09   
AABA     NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
AAC        NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
AAL      NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

        ADR TSO  IPO_Year            Sector  \
Symbol                                        
A           NaN    1999.0     Capital Goods   
AA          NaN    2016.0  Basic Industries   
AABA        NaN       NaN        Technology   
AAC         NaN    2014.0       Health Care   
AAL         NaN       NaN    Transportation   

                                                Industry  
Symbol                                                    
A       Biotechnology: Laboratory Analytical Instruments  
AA                                              Aluminum  
AABA                                        EDP Services  
AAC                                 Medical Specialities  
AAL                        Air Freight/Delivery Services

In [11]:
# run the cell! (create RangeIndex)
listings.reset_index(inplace = True)

In [12]:
# run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price    Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
1     AA     NYSE              Alcoa Corporation       29.15  5.407810e+09   
2   AABA   NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
3    AAC     NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

   ADR TSO  IPO_Year            Sector  \
0      NaN    1999.0     Capital Goods   
1      NaN    2016.0  Basic Industries   
2      NaN       NaN        Technology   
3      NaN    2014.0       Health Care   
4      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

43. Check the number of __unique values__ in __all columns__ of the __listings__ DataFrame! __How many__ unique values are in the __"Exchange"__ column?

In [13]:
listings.nunique()

Symbol        6840
Exchange         3
Name          5766
Last_Price    4470
Market_Cap    5835
ADR TSO        140
IPO_Year        42
Sector          12
Industry       135
dtype: int64

There are ... 3 unique values in the "Exchange" column!

44. __Select__ the __10__ listings with the __highest Market_Cap__! What is the Company __Name__ of the __third largest__ listing?

In [14]:
listings.nlargest(n = 10, columns = "Market_Cap")

Symbol Exchange                           Name  Last_Price    Market_Cap  \
11     AAPL   NASDAQ                     Apple Inc.      200.62  9.459795e+11   
4109   MSFT   NASDAQ          Microsoft Corporation      120.19  9.221233e+11   
350    AMZN   NASDAQ               Amazon.com, Inc.     1847.33  9.074138e+11   
2767  GOOGL   NASDAQ                  Alphabet Inc.     1206.45  8.387076e+11   
2766   GOOG   NASDAQ                  Alphabet Inc.     1202.16  8.357253e+11   
2262     FB   NASDAQ                 Facebook, Inc.      177.82  5.074965e+11   
623    BABA     NYSE  Alibaba Group Holding Limited      186.19  4.826388e+11   
3434    JNJ     NYSE              Johnson & Johnson      135.58  3.610186e+11   
6412      V     NYSE                      Visa Inc.      158.56  3.590806e+11   
6770    XOM     NYSE        Exxon Mobil Corporation       81.56  3.453905e+11   

      ADR TSO  IPO_Year             Sector  \
11        NaN    1980.0         Technology   
4109      NaN    1986.0         Technology   
350       NaN    1997.0  Consumer Services   
2767      NaN       NaN         Technology   
2766      NaN    2004.0         Technology   
2262      NaN    2012.0         Technology   
623       NaN    2014.0      Miscellaneous   
3434      NaN       NaN        Health Care   
6412      NaN       NaN      Miscellaneous   
6770      NaN       NaN             Energy   

                                             Industry  
11                             Computer Manufacturing  
4109          Computer Software: Prepackaged Software  
350                    Catalog/Specialty Distribution  
2767  Computer Software: Programming, Data Processing  
2766  Computer Software: Programming, Data Processing  
2262  Computer Software: Programming, Data Processing  
623                                 Business Services  
3434                            Major Pharmaceuticals  
6412                                Business Services  
6770                         Integrated oil Companies

The third largest company in terms of Market Capitalization is ... Amazon!

45. __Select__ the __15__ listings with the __lowest Prices__ (Penny Stocks)! Which Company has the __fourth lowest__ Stock Price?

In [15]:
listings.nsmallest(n = 15, columns = "Last_Price")

Symbol Exchange                                        Name  Last_Price  \
1676  CYTXZ   NASDAQ                    Cytori Therapeutics Inc.      0.0030   
3375  JASNW   NASDAQ                      Jason Industries, Inc.      0.0040   
6439  VEACW   NASDAQ            Vantage Energy Acquisition Corp.      0.0050   
6650  WHLRW   NASDAQ  Wheeler Real Estate Investment Trust, Inc.      0.0070   
5684  SNOAW   NASDAQ                Sonoma Pharmaceuticals, Inc.      0.0100   
3718  LMFAW   NASDAQ                    LM Funding America, Inc.      0.0115   
387   APDNW   NASDAQ                    Applied DNA Sciences Inc      0.0151   
337   AMRWW   NASDAQ                   Alta Mesa Resources, Inc.      0.0207   
2239  EYEGW   NASDAQ               Eyegate Pharmaceuticals, Inc.      0.0213   
1844  DRIOW   NASDAQ                           DarioHealth Corp.      0.0230   
5560  SHIPW   NASDAQ             Seanergy Maritime Holdings Corp      0.0270   
1177  CETXW   NASDAQ                                Cemtrex Inc.      0.0300   
4481  NUROW   NASDAQ                           NeuroMetrix, Inc.      0.0354   
4864  PHIOW   NASDAQ                  Phio Pharmaceuticals Corp.      0.0388   
4693  OXBRW   NASDAQ                Oxbridge Re Holdings Limited      0.0401   

      Market_Cap  ADR TSO  IPO_Year             Sector  \
1676         NaN      NaN       NaN        Health Care   
3375         NaN      NaN    2013.0  Consumer Durables   
6439         NaN      NaN    2017.0            Finance   
6650         NaN      NaN       NaN  Consumer Services   
5684         NaN      NaN       NaN        Health Care   
3718         NaN      NaN    2015.0            Finance   
387          NaN      NaN       NaN  Consumer Services   
337          NaN      NaN    2017.0             Energy   
2239         NaN      NaN       NaN        Health Care   
1844         NaN      NaN       NaN        Health Care   
5560         NaN      NaN       NaN     Transportation   
1177         NaN      NaN       NaN      Capital Goods   
4481         NaN      NaN       NaN        Health Care   
4864         NaN      NaN       NaN        Health Care   
4693         NaN      NaN    2014.0            Finance   

                                    Industry  
1676              Medical/Dental Instruments  
3375  Miscellaneous manufacturing industries  
6439                       Business Services  
6650           Real Estate Investment Trusts  
5684                   Major Pharmaceuticals  
3718              Finance: Consumer Services  
387                  Other Consumer Services  
337                     Oil & Gas Production  
2239                   Major Pharmaceuticals  
1844              Medical/Dental Instruments  
5560                   Marine Transportation  
1177         Industrial Machinery/Components  
4481              Medical/Dental Instruments  
4864                   Major Pharmaceuticals  
4693              Property-Casualty Insurers

The fourth lowest Stock Price has ... Wheeler Real Estate Investment Trust!

In [16]:
# run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price    Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
1     AA     NYSE              Alcoa Corporation       29.15  5.407810e+09   
2   AABA   NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
3    AAC     NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

   ADR TSO  IPO_Year            Sector  \
0      NaN    1999.0     Capital Goods   
1      NaN    2016.0  Basic Industries   
2      NaN       NaN        Technology   
3      NaN    2014.0       Health Care   
4      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [17]:
#run the cell!
listings.tail()

Symbol Exchange                           Name  Last_Price    Market_Cap  \
6847   ZUMZ   NASDAQ                    Zumiez Inc.       26.72  6.819191e+08   
6848    ZUO     NYSE                    Zuora, Inc.       19.79  2.147266e+09   
6849   ZYME     NYSE                 Zymeworks Inc.       15.74  5.040782e+08   
6850   ZYNE   NASDAQ  Zynerba Pharmaceuticals, Inc.        7.85  1.653995e+08   
6851   ZYXI   NASDAQ                    Zynex, Inc.        5.02  1.618349e+08   

      ADR TSO  IPO_Year             Sector  \
6847      NaN    2005.0  Consumer Services   
6848      NaN    2018.0         Technology   
6849      NaN    2017.0        Health Care   
6850      NaN    2015.0        Health Care   
6851      NaN       NaN        Health Care   

                                               Industry  
6847                     Clothing/Shoe/Accessory Stores  
6848            Computer Software: Prepackaged Software  
6849                              Major Pharmaceuticals  
6850                              Major Pharmaceuticals  
6851  Biotechnology: Electromedical & Electrotherape...

46. __Check__ for all elements in the column __"Exchange"__, whether they are __equal to "NASDAQ"__. __Save__ the boolean Series in the variable __mask1__! <br>
__Fill in the gaps__!

In [18]:
mask1 = listings.Exchange == "NASDAQ"
mask1

0       False
1       False
2        True
3       False
4        True
5       False
6        True
7       False
8        True
9        True
10      False
11       True
12      False
13      False
14       True
15       True
16       True
17      False
18      False
19      False
20      False
21       True
22       True
23       True
24       True
25      False
26      False
27       True
28       True
29      False
        ...  
6822     True
6823    False
6824     True
6825    False
6826     True
6827     True
6828     True
6829     True
6830     True
6831     True
6832     True
6833     True
6834     True
6835     True
6836     True
6837     True
6838    False
6839     True
6840    False
6841     True
6842     True
6843     True
6844    False
6845    False
6846    False
6847     True
6848    False
6849    False
6850     True
6851     True
Name: Exchange, Length: 6852, dtype: bool

47. __Check__ for all elements in the column __"Last_Price"__, whether they are __less than 1.0__. __Save__ the boolean Series in the variable __mask2__!

In [19]:
mask2 = listings.Last_Price < 1

48. __Filter__ the DataFrame __listings__ for all __NASDAQ-listings__ (mask1) and __save__ the resulting DataFrame in the variable __nasdaq__ (create a copy)! <br>
__Fill in the gaps__!

In [20]:
nasdaq = listings.loc[mask1].copy()

__How many NASDAQ-listings__ does the __nasdaq__ DataFrame contain?

In [21]:
# run the cell!
nasdaq.head()

Symbol Exchange                           Name  Last_Price    Market_Cap  \
2   AABA   NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   
6   AAME   NASDAQ  Atlantic American Corporation        2.48  4.998398e+07   
8   AAOI   NASDAQ  Applied Optoelectronics, Inc.       12.30  2.445561e+08   
9   AAON   NASDAQ                     AAON, Inc.       44.88  2.336955e+09   

   ADR TSO  IPO_Year          Sector                         Industry  
2      NaN       NaN      Technology                     EDP Services  
4      NaN       NaN  Transportation    Air Freight/Delivery Services  
6      NaN       NaN         Finance                   Life Insurance  
8      NaN    2013.0      Technology                   Semiconductors  
9      NaN       NaN   Capital Goods  Industrial Machinery/Components

In [22]:
# run the cell!
nasdaq.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3435 entries, 2 to 6851
Data columns (total 9 columns):
Symbol        3435 non-null object
Exchange      3435 non-null object
Name          3435 non-null object
Last_Price    3432 non-null float64
Market_Cap    3168 non-null float64
ADR TSO       140 non-null float64
IPO_Year      1599 non-null float64
Sector        2964 non-null object
Industry      2964 non-null object
dtypes: float64(4), object(5)
memory usage: 268.4+ KB


There are in total ... 3435 NASDAQ-listings!

49. __Filter__ the DataFrame __listings__ for all __Penny Stocks__ (mask2) and __save__ the resulting DataFrame in the variable __penny__ (create a copy)!

In [23]:
penny = listings.loc[mask2].copy()

__How many Penny Stocks__ does the __penny__ DataFrame contain?

In [24]:
# run the cell!
penny.head()

Symbol Exchange                        Name  Last_Price    Market_Cap  \
13     AAU     AMEX      Almaden Minerals, Ltd.      0.5750  6.424286e+07   
67    ACST   NASDAQ         Acasti Pharma, Inc.      0.9500  7.422610e+07   
81   ADILW   NASDAQ  Adial Pharmaceuticals, Inc      0.3705           NaN   
87    ADOM   NASDAQ               ADOMANI, Inc.      0.3426  2.495277e+07   
115   AEMD   NASDAQ       Aethlon Medical, Inc.      0.8300  1.573722e+07   

     ADR TSO  IPO_Year            Sector  \
13       NaN    2015.0  Basic Industries   
67       NaN       NaN       Health Care   
81       NaN    2018.0       Health Care   
87       NaN    2017.0     Capital Goods   
115      NaN       NaN     Capital Goods   

                                             Industry  
13                                    Precious Metals  
67                              Major Pharmaceuticals  
81                              Major Pharmaceuticals  
87                                  Auto Parts:O.E.M.  
115  Biotechnology: Laboratory Analytical Instruments

In [25]:
# run the cell!
penny.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 327 entries, 13 to 6840
Data columns (total 9 columns):
Symbol        327 non-null object
Exchange      327 non-null object
Name          327 non-null object
Last_Price    327 non-null float64
Market_Cap    199 non-null float64
ADR TSO       6 non-null float64
IPO_Year      161 non-null float64
Sector        324 non-null object
Industry      324 non-null object
dtypes: float64(4), object(5)
memory usage: 25.5+ KB


There are in total ... 327 Penny Stocks!

50. __Filter__ the DataFrame __listings__ for all __Penny Stocks__ (mask2) that are __NASDAQ-listed__ (mask1) and __save__ the resulting DataFrame in the variable __nasdaq_penny__! __Fill in the gaps!__

In [26]:
nasdaq_penny = listings.loc[mask1 & mask2]

__How many listings__ does the __nasdaq_penny__ DataFrame contain?

In [27]:
# run the cell!
nasdaq_penny.head()

Symbol Exchange                                  Name  Last_Price  \
67    ACST   NASDAQ                   Acasti Pharma, Inc.      0.9500   
81   ADILW   NASDAQ            Adial Pharmaceuticals, Inc      0.3705   
87    ADOM   NASDAQ                         ADOMANI, Inc.      0.3426   
115   AEMD   NASDAQ                 Aethlon Medical, Inc.      0.8300   
122   AETI   NASDAQ  American Electric Technologies, Inc.      0.8741   

       Market_Cap  ADR TSO  IPO_Year         Sector  \
67   7.422610e+07      NaN       NaN    Health Care   
81            NaN      NaN    2018.0    Health Care   
87   2.495277e+07      NaN    2017.0  Capital Goods   
115  1.573722e+07      NaN       NaN  Capital Goods   
122  7.927501e+06      NaN       NaN         Energy   

                                             Industry  
67                              Major Pharmaceuticals  
81                              Major Pharmaceuticals  
87                                  Auto Parts:O.E.M.  
115  Biotechnology: Laboratory Analytical Instruments  
122                   Industrial Machinery/Components

In [28]:
# run the cell!
nasdaq_penny.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 257 entries, 67 to 6836
Data columns (total 9 columns):
Symbol        257 non-null object
Exchange      257 non-null object
Name          257 non-null object
Last_Price    257 non-null float64
Market_Cap    131 non-null float64
ADR TSO       6 non-null float64
IPO_Year      144 non-null float64
Sector        256 non-null object
Industry      256 non-null object
dtypes: float64(4), object(5)
memory usage: 20.1+ KB


There are in total ... 257 penny stocks listed on nasdaq!

# Well Done!

------------------------------

# Hints (Spoiler!)

39. sort_values() method, by parameter

40. pass lists to the by and ascending parameter

41. set_index() method, inplace

42. sort_index() method, inplace

43. nunique() method

44. nlargest() method, parameters n and columns

45. nsmallest() method

46. listings.--- == "---"

47. listings.--- < ---

48. pass mask1 to the loc operator

49. listings.loc[---].---

50. listings.loc[--- & ---]